### DISCUSSION TITLE
RNode on ESP32C3, freeze during modem init

### DISCUSSION THREAD

**Discussion by nathmo** [markqvist/Reticulum#529] [Help & Questions]
Hi,
I'm looking for help, tips, pointer on how to debug the following issue : 
I'm trying to port the RNode firmware on the ESP32-C3.

here is my issue with my current progress : https://github.com/liberatedsystems/RNode_Firmware_CE/issues/18

the TLDR is that the node freeze after the following happen : 
when the software try to initialise the radio modem, the SPI beginTransaction function make the ESP32-C3 freeze.

using print statement, I traced the issue back to theses lines : 

1) in the RNode_Firmware_CE.ino : Line 230 if (selected_radio->preInit()) 
2) in Radio.cpp : Line 1138 of sx127x::preInit() version = readRegister(REG_VERSION_7X);
3) in Radio.cpp : Line 1098 of sx127x::readRegister(uint8_t address) { return singleTransfer(address & 0x7f, 0x00); }
4) in Radio.cpp : Line 1158 of sx127x::singleTransfer : _spiModem.beginTransaction(_spiSettings)

@jacobeva seems to believe it might be due to a stack issues. I tried to set a bigger stack as his fix suggested, but there is no behavior change. 

I also tried to disable the SPI init (https://github.com/nathmo/RNode_Firmware_CE/commit/2217fe07714017086a672b2e19b34d6271e378c3) which allowed the board to boot and respond to command but without much surprise, the radio don't work. 

I believe it's a semaphore issue, but I don't know how to debug these kinds of issue. Especially since I don't see anything in the code that was run before that would have touched the SPI.
(assert failed: xQueueSemaphoreTake queue.c:1549 (pxQueue->uxItemSize == 0))

If someone with more experience than me could spend some time to look at it, I would be incredibly grateful :)

this is my version : https://github.com/nathmo/RNode_Firmware_CE

This is the crashdump I got : 
```
Rebooting...
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0xe (SPI_FAST_FLASH_BOOT)
Saved PC:0x403823a0
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5810,len:0x438
load:0x403cc710,len:0x90c
load:0x403ce710,len:0x2624
entry 0x403cc710
pre init before begin                                        // my debug print statement
_spiModem.begin(_sclk, _miso, _mosi, _ss)      // my debug print statement
pre init after begin                                          // my debug print statement
A                                                                      // my debug print statement

assert failed: xQueueSemaphoreTake queue.c:1549 (pxQueue->uxItemSize == 0)
Core  0 register dump:
MEPC    : 0x4038279a  RA      : 0x4038a5be  SP      : 0x3fca4d00  GP      : 0x3fc91400
TP      : 0x3fc7f054  T0      : 0x37363534  T1      : 0x7271706f  T2      : 0x33323130
S0/FP   : 0x3fca4e7a  S1      : 0x0000007d  A0      : 0x3fca4d64  A1      : 0x3fc930f9
A2      : 0x00000001  A3      : 0x00000029  A4      : 0x00000001  A5      : 0x3fc98000
A6      : 0x7a797877  A7      : 0x76757473  S2      : 0x3fca4d58  S3      : 0x00000001
S4      : 0x3fca4d58  S5      : 0x4038b25c  S6      : 0x00000000  S7      : 0x00000000
S8      : 0x00000000  S9      : 0x00000000  S10     : 0x00000000  S11     : 0x00000000
T3      : 0x6e6d6c6b  T4      : 0x6a696867  T5      : 0x66656463  T6      : 0x62613938
MSTATUS : 0x00001801  MTVEC   : 0x40380001  MCAUSE  : 0x00000007  MTVAL   : 0x00000000
MHARTID : 0x00000000

Stack memory:
3fca4d00: 0x6e6d6c6b 0x6a696867 0x66656463 0x4038f74c 0x3fc9f1a4 0x3fca0f94 0x000001c0 0x3fc930f8
3fca4d20: 0x00000020 0x00000017 0x00000000 0x39343531 0xa5a5a500 0x3fc931e0 0x3c0c78d4 0x3fc935bc
3fca4d40: 0x3c0c797d 0x3fc931f0 0x3fca4d2c 0x3fc931f4 0x3c0c7d18 0x3fc930f8 0x00000000 0x00000000
3fca4d60: 0xff000000 0x65737361 0x66207472 0x656c6961 0x78203a64 0x75657551 0x6d655365 0x6f687061
3fca4d80: 0x61546572 0x7120656b 0x65756575 0x313a632e 0x20393435 0x51787028 0x65756575 0x78753e2d
3fca4da0: 0x6d657449 0x657a6953 0x203d3d20 0x00002930 0x00000002 0x3fc96000 0x00000001 0x00000064
3fca4dc0: 0x00000003 0x00000004 0x600c2000 0x600c2000 0x00000000 0x00000000 0x3fc9f490 0x4038ad0e
3fca4de0: 0x00000000 0x00000042 0x00000000 0x4038c632 0x3fc977f0 0x00000000 0x00000000 0x00000000
3fca4e00: 0x3c0c0514 0x3fc9f43c 0x3fc9f39c 0x4038a80a 0x00000000 0x00000000 0x3fc9f39c 0x4038ad28
3fca4e20: 0x00000000 0x00000017 0x00000000 0x48587a27 0x00000000 0x00000042 0x00000000 0x000012c8
3fca4e40: 0x00000001 0x007a1200 0x3fc9d7c8 0x4038b25c 0x60043000 0x00000003 0x00000000 0xffffffff
3fca4e60: 0x3fc90c40 0x3fc977f8 0x00000017 0x3fc977e0 0x60043000 0x00000017 0x00000000 0x000012c8
3fca4e80: 0x00000001 0x007a1200 0x3fca9418 0x42005bb8 0x00000000 0x00000042 0x00000000 0x30b80001
3fca4ea0: 0x3fca9150 0x3fc96000 0x3fca9418 0x403803ba 0x00000000 0x00000000 0x00000012 0x3fca8150
3fca4ec0: 0x000001f3 0x3fc96000 0x00001432 0x42004bc2 0x00000005 0x3fc94200 0x00000000 0x00000000
3fca4ee0: 0x00002000 0x3fc977d0 0x3fca8150 0x420029c4 0x00000002 0xffffffff 0x00000000 0x00000000
3fca4f00: 0x00000000 0x00000000 0xffffff00 0x000000ff 0x00000000 0x000f4240 0x00000000 0x3fc9d7c8
3fca4f20: 0x00000000 0x3fc98000 0x00000000 0x4200a0a6 0x00000000 0x00000000 0x00000000 0x4038cc4e
3fca4f40: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5
3fca4f60: 0xa5a5a5a5 0xa5a5a5a5 0xbaad5678 0x00000160 0xabba1234 0x00000154 0x3fca4c90 0x0000138c
3fca4f80: 0x3fc96c1c 0x3fc96c1c 0x3fca4f78 0x3fc96c14 0x00000018 0x10d890ce 0xd1fb350e 0x3fca4f78
3fca4fa0: 0x00000000 0x00000001 0x3fca1368 0x706f6f6c 0x6b736154 0x4dcc4500 0x002526d0 0x00000000
3fca4fc0: 0x3fca4f60 0x00000001 0x00000000 0x00000000 0x00000000 0x00000000 0x3fc9d064 0x3fc9d0cc
3fca4fe0: 0x3fc9d134 0x00000000 0x00000000 0x00000001 0x00000000 0x00000000 0x00000000 0x4209fc76
3fca5000: 0x00000000 0x00000000 0x00000000 0x3fc9f374 0x00000000 0x00000000 0x00000000 0x00000000
3fca5020: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
3fca5040: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
3fca5060: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
3fca5080: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
3fca50a0: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
3fca50c0: 0x00000000 0x00000000 0xad000000 0xbaad5678 0x00001860 0xabba1234 0x00001854 0x3fca5130
3fca50e0: 0x3fca5130 0x3fca6930 0x3fca692f 0x00000000 0x3fca50f4 0xffffffff 0x3fca50f4 0x3fca50f4



ELF file SHA256: 0426eedcb4542d9f

Rebooting...
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0xe (SPI_FAST_FLASH_BOOT)
Saved PC:0x403823a0
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5810,len:0x438
load:0x403cc710,len:0x90c
load:0x403ce710,len:0x2624
entry 0x403cc710
```

**Comment by cobraPA** [comment]
Greetings, could you share more info about the board?  Is it a custom board or something off the shelf?  Or would you rather not say?  What modem is installed that is trying to be run for the preInit?

**Reply by nathmo** [reply]
this is the picture from the github issue I mentionned earlier : (https://github.com/liberatedsystems/RNode_Firmware_CE/issues/18)
![image](https://github.com/user-attachments/assets/1d129790-cfe4-4e7d-993e-9356af41bf68)
its a ESP32-C3 Mini (https://fr.aliexpress.com/item/1005006211660074.html)
I hooked up a SX1278 module with Dupont wire. (https://fr.aliexpress.com/item/1005005968916380.html)

I checked the wiring (and made a test sketch to communicate with the LoRa board. It works) 


from what I understand of my test, the crash happen due to a semaphore on the SPI bus.
What lock that semaphore is that SPI object is passed by value (copy) and not by reference.
This creates a new SPI object, but it fails when I want to use it because the other one still exist and hold the lock.

sx127x::sx127x(uint8_t index, SPIClass spi,...
sx127x::sx127x(uint8_t index, SPIClass& spi, ...

I tried to change the code to pass it by reference instead but now the init must fail somehow since I can't  configure the LoRa modem

**Reply by nathmo** [reply]
and here is my code : https://github.com/nathmo/RNode_Firmware_CE

**Reply by cobraPA** [reply]
Ok, I do see that in the other thread..  Did you start out by turning off things you don't need right away in your board config?  And did you add a new one?  Things like HAS_CONSOLE, HAS_DISPLAY, HAS_BLUETOOTH or HAS_BLE?  That simplifies things because you don't have to worry about them breaking the board before you test them.  And CONSOLE will turn off the WIFI, no Bluetooth will block off that part of the MAC address (or you can just stub that out then in the code.)  It also makes the code smaller by leaving those code blocks out, in case you have an initial size issue with the firmware.

**Reply by nathmo** [reply]
I did in some test. It is not changing much as the modem initialization is among the very first thing that happen

**Reply by nathmo** [reply]
well we made progress. so far the board is booting and it was a semaphore issue. but now the LoRa modem is not switching to the right frequency when used with Reticulum. more test is required
